File: okj/rendszeruzemelteto/hianyzasok/python/Hianyzasok.py
from dataclasses import dataclass
from itertools import groupby

@dataclass
class Hianyzas:
    nev: str
    osztaly: str
    elsoNap: int
    utolsoNap: int
    mulasztottOrak: int


with open('szeptember.csv', encoding = 'utf-8') as file:
    create_hianyzas = lambda split: Hianyzas(split[0], split[1], int(split[2], int(split[3], int(split[4]))))
    hianyzasok = [ create_hianyzas(k.split(';')) for k in file.readlines()[1:] ]

totalHianyzottOrak = sum(k.mulasztottOrak for k in hianyzasok)

print(f'2. Feladat: Hianyzott orak: {totalHianyzottOrak}')

bekertNap = int(input('3. Feladat: Írj be egy napot(1-30)!'))
bekertNev = input('Írj be 1 nevet!')
bekert_hianyzott_e = any(True for k in hianyzasok if k.nev == bekertNev)

print(f'4. Feladat: {bekertNev} {"hiányzott" if bekert_hianyzott_e else "nem hiányzott"}')
print('5. Feladat')

azonANaponHianyoztak = [ k for k in hianyzasok if bekertNap >= k.elsoNap and bekertNap <= k.utolsoNap ]
if len(azonANaponHianyoztak) == 0:
    print('Nem volt hiányzó')
else:
    for hiany in azonANaponHianyoztak:
        print(f'{hiany.nev} {hiany.osztaly}')

hianyzasok.sort(key = lambda k: k.osztaly)

orakatSzamol = lambda m: sum(k.mulasztottOrak for k in m)
hianyzasokStat = dict((k, orakatSzamol(v)) for k, v in groupby(hianyzasok, key = lambda m: m.osztaly))

with open('osszesites.csv', 'w') as outFile:
    for osztaly, totalMulasztottOrak in hianyzasokStat.items():
        outFile.write(f'{osztaly};{totalMulasztottOrak}\n')

File: okj/rendszeruzemelteto/hianyzasok/rust/Hianyzasok.rs
use std::fs::File;
use std::io::{ BufReader, BufRead, Write, stdin };

use itertools::Itertools;

struct Hianyzas {
    nev: String,
    osztaly: String,
    elso_nap: i32,
    utolso_nap: i32,
    mulasztott_orak: i32
}

fn create_hianyzas(line: &String) -> Hianyzas {
    let split = line.split(';').collect::<Vec<&str>>();

    Hianyzas {
        nev: split[0].to_string(),
        osztaly: split[1].to_string(),
        elso_nap: split[2].parse::<i32>().unwrap(),
        utolso_nap: split[3].parse::<i32>().unwrap(),
        mulasztott_orak: split[4].parse::<i32>().unwrap()
    }
}

fn main() {
    let input_file = File::open("szeptember.csv").unwrap();
    let hianyzasok = BufReader::new(input_file)
                              .lines()
                              .skip(1)
                              .map(|k| create_hianyzas(&k.unwrap()))
                              .collect::<Vec<Hianyzas>>();

    let total_hianyzott_orak = hianyzasok.iter()
                                         .map(|k| k.mulasztott_orak)
                                         .sum::<i32>();

    println!("2. Feladat: Hiányzott órák: {}", total_hianyzott_orak);
    println!("{}", "3. Feladat: Írj be egy napot(1-30) és egy nevet!");

    let mut bekert_nap_str = String::new();
    let mut bekert_nev_str = String::new();
    stdin().read_line(&mut bekert_nap_str);
    stdin().read_line(&mut bekert_nev_str);

    let bekert_nap = bekert_nap_str.trim_end().parse::<i32>().unwrap();
    let bekert_nev = bekert_nev_str.trim_end();

    println!("{}", "4. Feladat:");

    let bekert_hianyzott_e = hianyzasok.iter()
                                       .any(|k| k.nev == bekert_nev);

    println!("4. Feladat: {} {}", bekert_nev, if bekert_hianyzott_e { "hiányzott" } else { "nem hiányzott" });
    println!("5. Feladat:");

    let azon_a_napon_hianyoztak = hianyzasok.iter()
                                            .filter(|k| bekert_nap >= k.elso_nap && bekert_nap <= k.utolso_nap)
                                            .collect::<Vec<&Hianyzas>>();

    if azon_a_napon_hianyoztak.len() == 0 {
        println!("Nem volt hiányzó");
    }else{
        azon_a_napon_hianyoztak.iter().for_each(|k| println!("{} {}", k.nev, k.osztaly));
    }

    let to_file = hianyzasok.iter()
                            .sorted_by_key(|k| &k.osztaly)
                            .group_by(|k| &k.osztaly)
                            .into_iter()
                            .map(|(osztaly, items)| format!("{};{}", osztaly, items.map(|k| k.mulasztott_orak).sum::<i32>()))
                            .join("\n");

    let mut output = File::create("osszesites.csv").unwrap();
    write!(output, "{}", to_file);
}